### PR TITLE
Fix loadPackage() to call mkdir only when needed

### LINKED ISSUE #4736: loadPackage() not to access the file system if the cache already exists
## 🚀 Feature

`pyodide.loadPackage()` should not access the file system with `nodeFsPromisesMod.mkdir` when not necessary, or the cache dir already exists.
It can be done automatically, or by providing an option to skip the file system access.

### Motivation

`loadPackage()` [calls `nodeFsPromisesMod.mkdir` any time](https://github.com/pyodide/pyodide/blob/c4a6964429fe4904c33a6a6b5771b173bf4ee9f8/src/js/load-package.ts#L261-L263) to ensure the directory exists.
However, it causes an error if the file system is read-only even when the directory and the cached packages already exist.
My actual use case is to ship Pyodide in an Electron app with the downloaded package files. In this case, the bundled files are visible to the Node process of the Electron app via `fs` module but not writable as they are packaged in an [ASAR archive](https://www.electronjs.org/docs/latest/tutorial/asar-archives) and [the archive is read-only](https://www.electronjs.org/docs/latest/tutorial/asar-archives#archives-are-read-only). It's https://github.com/whitphx/stlite/issues/831#issuecomment-2081755934.

### Pitch

* One option is to call `mkdir` only when the cache directory does not exist.
* Another is to add an option to skip the file system access, e.g. `loadPackage(packageName, { skipDownload: true })`.

### PR DESCRIPTION
**Pull request by whitphx**
### Description

Resolves #4736

### Checklists

- [x] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry
- [ ] Add / update tests
- [ ] Add new / update outdated documentation

### CODE REVIEW

**Comment by hoodmane** [comment]
Thanks for working on this @whitphx!

File: src/js/load-package.ts [hunk 1]
```diff
@@ -257,10 +257,13 @@ async function downloadPackage(
...
-    await nodeFsPromisesMod.mkdir(API.config.packageCacheDir, {
-      recursive: true,
-    });
+    // Ensure that the directory exists before trying to download files into it.
+    // Call mkdir() only when the dir doesn't exist to avoid an error on read-only file systems. See https://github.com/pyodide/pyodide/issues/4736
+    await nodeFsPromisesMod.access(API.config.packageCacheDir).catch(() =>
+      nodeFsPromisesMod.mkdir(API.config.packageCacheDir, {
+        recursive: true,
+      }),
+    );
```
Comment (hoodmane) [inline src/js/load-package.ts:266]: We prefer standard control flow to callbacks:
```js
try {
   // Check if packageCacheDir exists
   await nodeFsPromisesMod.access(API.config.packageCacheDir);
} catch(e) {
   // It doesn't exist, make it.
   await nodeFsPromisesMod.mkdir(API.config.packageCacheDir, {
      recursive: true,
   }),
}

**Review by hoodmane — APPROVED** [review]
Do we want some sort of test for this? It can merge without but it would be nice to add. Please add a changelog entry. Thanks @whitphx!

**Comment by whitphx** [comment]
Thank you!
Will try to add a test.

**Comment by hoodmane** [comment]
Okay. Let me know if you give up on adding a test and if so we can merge as-is.

**Comment by whitphx** [comment]
@hoodmane Thanks, now let me give up it as I couldn't find a way to assert file system access or spying `fs.mkdir` with the current setup.

**Comment by ryanking13** [comment]
Thanks @whitphx!

**Pull request merged**